### PR TITLE
For Discussion: pylsl migrated from ctypes to cffi. 

### PR DESCRIPTION
**Pull request by cboulay**
Also added support for post-processing flags on a stream and passing in pushing/pulling objects that implement Python buffers.

Please do not merge this yet. At this stage, I am making this pull request for discussion purposes only.

### CODE REVIEW

**Comment by cboulay** [comment]
This is an attempt to migrate from using ctypes to using cffi. There are several advantages to cffi that I know of.

1. The interface is generated by parsing regular c-code. This can be done using the lsl_c.h file (with a tiny change to make it easier to snip out things the parser doesn't recognize). The result is c-compiled and linked to pylsl.dll. This makes for robust addressing of pylsl.py and eliminates the need for ctypes argtypes and restypes (and the segfaults we were getting with those).
1. It seems to be well-supported and continuously developed by the PyPy folks, while ctypes support seems to be waning.
1. Arguments with 'simple' data types (not pointers, buffers) can be passed to the library without any c-type conversion.
1. The support for dealing with data buffers seems to be pretty straight-forward so far.

Some disadvantages
1. It requires a c-compiler on the user's machine to install. The c-compiler isn't strictly necessary for using cffi to wrap the library; see [this note](https://github.com/sccn/labstreaminglayer/pull/124/files#diff-5e254b034f50198b859b4b5b04dd6ba3R67) about how to byte-compile, and the pylsl OO interface can fall back to loading the byte-compiled version instead of the c-compiled version. Unfortunately, I can't figure out how to get setuptools to fall back to byte-compiling instead of c-compiling during install.
1. The way the linker/installer handles library rpath's is a little strange to me. Everytime I `python3 setup.py install` I have to follow that up with a `site-packages $install_name_tool -change @rpath/liblsl64.dylib @loader_path/liblsl64.dylib pylsl-1.10.4-py3.5-macosx-10.11-x86_64.egg/pylsl/_pylsl.cpython-35m-darwin.so` I have a [question pending in cffi's mailing list](https://groups.google.com/forum/#!topic/python-cffi/IZNTdGZy3ko).

**Comment by cboulay** [comment]
Prior to these changes, and indeed one of the major motivations for these changes, was that I was getting performance hitches in my app. Most processing iterations took 10-12 msec, but every 4th iteration jumped up to 60 msec. (Each iteration starts at the call to the inlet and ends when the data are sent for visualization, with some filtering, segmentation, frequency transforms in between). On profiling, I seemed to spend a lot of time in pylsl.

Just switching over to cffi from ctypes didn't give me much of a performance boost or even smooth out the hitches I was having. In fact, it may have made my performance slightly worse on the fast frames (+2-3 msec) which may have slightly reduced the hitching on the 4th frame. Until the disadvantages from above get sorted out, or we decide to forego the c-compiling (but lose some of the robustness in doing so), it might not be worth switching. Or maybe it still is simply because the absence of segfaults in OS X.

I made two other significant changes in this pull request, both of which could be accomplished using ctypes, I think.

First, I added in support for the post-processing flags on an inlet. This gave me a small performance increase vs calling inlet.time_correction() but not much. The greater increase was because we no longer need a separate python processing step to dejitter and force monotonicity.

Second, I added support for passing python objects that implement python buffer. When using this feature, no copy operations are required and the list comprehension to change the pull_chunk result to a list of lists is not needed. Similar speedups are possible when pushing. I mentioned this at the bottom of #108 for numpy arrays and ctypes. cffi makes this easy to do with arbitrary buffered objects, but this might also be true in ctypes.

With these two changes, I no longer get hitches in my processing pipeline. Every iteration is between 10-14 msecs.

**Comment by cboulay** [comment]
Still incomplete in this PR:

I need a way to automatically copy lsl_c.h from pylsl/include into the liblsl-Python/pylsl dir prior to packaging for distribution. I'd appreciate any advice from anyone that knows setuptools.

**Comment by chkothe** [comment]
Hi Chad, good stuff -- much appreciated! Will review at next opportunity.

**Comment by chkothe** [comment]
Sorry about still not having gotten back to this, Chad -- currently in a crunch over some upcoming hackathon(s). I do like the prospect of a guaranteed resolution of ctypes marshaling issues, that would be my main impetus to merge that one in. The flavor that relies on a C compiler would be much less appealing to me than the pure-python one, though (as on non-UNIX platforms C compilers are traditionally not so easy to get hold of), but luckily it CFFI seems to support that ("ABI mode" if I recall correctly from my quick glance over the docs). 

Now coincidentally yesterday someone in the office managed to replicate a segfault with the current pylsl repeatedly, and in the process of debugging it I found a definitive marshaling bug in the current pylsl (due to the same wrong assumption that I originally had about the behavior of .restype; I actually fired off a bug [report](https://bugs.python.org/issue27871) on the ctypes docs regarding that now). I did a scan through the rest of the code and I believe that this was the last problem of that sort (a wonder that we didn't catch it earlier), so I now have a 1-line fix to pylsl, and maybe we can merge that with the postprocessing extension that you did and the data transfer optimization if it's straightforward before we make a final decision on the cffi topic. I'd like to make a PR for that fix and get it in asap, since I don't want this to crash on anyone.

**Comment by chkothe** [comment]
By the way, I figured that one doesn't actually need to specify the argtypes (if you pass in arguments with c_*** types then it'll be marshalled correctly), which reduces the pain level to just getting declaring the restypes properly; the issue was just that the handle in the XMLElement class was not of the c_void_p type that I thought it was, but int (see bug report).

**Comment by cboulay** [comment]
My reading of the reply to the bug report suggests that the object handle (`self.obj`) should not be a ctypes-type, nor should lib function arguments be wrapped with c_*** types. Instead, the argtypes should be specified and the arguments will be converted implicitly. But we [were specifying it](https://github.com/sccn/labstreaminglayer/blob/master/LSL/liblsl-Python/pylsl/pylsl.py#L1177-L1218), weren't we? When you were debugging that issue, was the function lacking an argtypes specification? And why would this only present in OSX?
(And it would also seem that specifying the restype type does nothing if it is a fundamental datatype. I agree this is worth a more explicit mention in the docs.)

Whether we decide to c_***() wrap arguments and not specify argtypes, or not wrap arguments and use argtypes (may not work properly), we should probably be consistent for all arguments. Presently there is a mix. Is there significant savings to be had by not doing a c_void_p(self.obj) inside every library function call?

I will branch off your other pull request and work on the other additions (post-processing extension, passing around pointers to buffers) there. I think it's going to be a little more work to get ctypes to work with arbitrary data buffers than it was for cffi. For pulling samples/chunks, I assume we don't want numpy as a dependency nor do we want flags for things like `returnType`, so I think we have to have the user pass in the destination object's buffer as an argument. Is there a different pattern you'd prefer?

File: LSL/liblsl-Python/pylsl/pylsl.py [hunk 1]
```diff
@@ -76,6 +107,13 @@
 # Can not be transmitted.
 cf_undefined = 0
 
+proc_none = 0
+proc_clocksync = 1
+proc_dejitter = 2
+proc_monotonize = 4
+proc_threadsafe = 5
```
Comment (cboulay) [inline LSL/liblsl-Python/pylsl/pylsl.py]: Probably should be `proc_threadsafe = 8`

**Comment by chkothe** [comment]
Oops :)

**Comment by cboulay** [comment]
Added this to the archive: https://github.com/labstreaminglayer/lsl_archived/pull/1

**Pull request closed without merging**